File: src/main/resources/Expansions/CustomSection.js
// 文件名不重要, 写成啥都行
// enable函数会自动执行
function enable() {
    // 导入相应的类, 看不懂的话直接抄就行
    const SectionManager = Packages.pers.neige.neigeitems.manager.SectionManager.INSTANCE
    const CustomSection = Packages.pers.neige.neigeitems.section.impl.CustomSection

    // 节点注册
    SectionManager.loadParser(
        new CustomSection(
            // 节点id
            "test",
            /**
             * 用于私有节点解析
             * @param data ConfigurationSection 节点内容
             * @param cache HashMap<String, String>? 解析值缓存
             * @param player OfflinePlayer? 待解析玩家
             * @param sections ConfigurationSection? 节点池
             * @return 解析值
             */
            function(data, cache, player, sections) {
                if (data.contains("values")) {
                    // SectionUtils.parseSection("待解析字符串", cache, player, sections)用于解析节点内容
                    return SectionUtils.parseSection("<number::0_1_2>", cache, player, sections)
                }
                return null
            },
            /**
             * 用于即时节点解析
             * @param args List<String> 节点参数
             * @param cache HashMap<String, String>? 解析值缓存
             * @param player OfflinePlayer? 待解析玩家
             * @param sections ConfigurationSection? 节点池
             * @return 解析值
             */
            function(args, cache, player, sections) {
                return SectionUtils.parseSection("<number::0_1_2>", cache, player, sections)
            }
        )
    )
}

File: project/common/src/main/kotlin/pers/neige/neigeitems/section/impl/CustomSection.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser

/**
 * 自定义节点, 具体将在js文件中实现, nashorn会自动解析匹配类型, 因此不用担心重定向问题, 不用将kotlin的function更改为java的function
 * 别问我这个类为什么不叫CustomParser, 等我反应过来的时候为时已晚
 */
class CustomSection(
    // 自定义节点ID
    override val id: String,

    // 用于响应私有节点解析
    private val func1: (
        ConfigurationSection,
        MutableMap<String, String>?,
        OfflinePlayer?,
        ConfigurationSection?
    ) -> String?,

    // 用于响应即时节点解析
    private val func2: (
        List<String>,
        MutableMap<String, String>?,
        OfflinePlayer?,
        ConfigurationSection?
    ) -> String?

) : SectionParser() {

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?): String? {
        return func1(data, cache, player, sections)
    }

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?): String {
        return func2(args, cache, player, sections) ?: "<$id::${args.joinToString("_")}>"
    }
}
